File: python/2022 ICPC Sinchon Summer/enew.py
import sys
from collections import deque
sys.setrecursionlimit(10**6)

input = lambda: sys.stdin.readline().rstrip()
pair = lambda : map(int, input().split())

n = int(input())
s = input()
tree = [[] for _ in range(n+1)]
visited = [0] * (n+1)
visited[1] = 1
dq = deque()
dq.append(1)
cand = []
res = []
max_depth = 1

for _ in range(n-1):
    a, b = pair()
    tree[a].append(b)
    tree[b].append(a)

for i in tree:
    i.sort()

def dfs(v, depth, max_depth):
    for i in tree[v]:
        if not visited[i]:
            visited[i] = 1
            dq.append(i)
            dfs(i, depth+1, max_depth)
            max_depth = depth + 1
            dq.pop()
    if depth >= max_depth:
        cand.append(list(dq))

dfs(1,1, max_depth)

for i in cand:
    tmp = ""
    for idx in i:
        tmp += s[idx-1]
    res.append(tmp)

res.sort()
print(res[-1])

File: python/2022 ICPC Sinchon Summer/enew.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;
#define MAX 5000001

int n;
string s;
bool visited[MAX];
vector<int> dq;
vector<vector<int>> tree(MAX);
vector<vector<int>> cand;
vector<string> res;
int max_depth = 1;

void dfs(int v, int depth, int map_depth) {
    for (auto &i : tree[v]) {
        if (visited[i] == false) {
            visited[i] = 1;
            dq.push_back(i);
            dfs(i,depth+1, max_depth);
            max_depth = depth + 1;
            dq.pop_back();
        }
    }
    if (depth >= max_depth) {
        vector<int> tmp(dq);
        cand.push_back(dq);
    }
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    cin >> n;
    cin >> s;
    visited[1] = true;
    dq.push_back(1);

    for (int i = 0; i < n-1; i++) {
        int a, b;
        cin >> a >> b;
        tree[a].push_back(b);
        tree[b].push_back(a);
    }

    for (auto &i : tree)
        sort(i.begin(), i.end());

    dfs(1, 1, max_depth);

    for (auto& i : cand) {
        string tmp;
        for (auto& j : i) {
            tmp += s[j-1];
        }
        res.push_back(tmp);
    }

    sort(res.begin(), res.end(), greater<string>());
    cout << res[0];
}